Add GDB support

lyakh opened the pull request:
Taking over https://github.com/nklayman/linux/commit/930457fc8f27a07e87276cbdcf26f4378ff83280 . Respective Zephyr and SOF PRs to follow next week.

nklayman committed:
ASoC: SOF: sof-client: expose Zephyr GDB stub

Adds a DebugFS file to communicate with Zephyr's GDB stub. Communication
is via ringbuffers in shared SRAM. Both IPC3 and IPC4 are supported.

Signed-off-by: Noah Klayman <noah.klayman@intel.com>
Co-developed-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+	unsigned char tail;
+	unsigned char fill2[63];
+	unsigned char data[RING_SIZE];
+} __packed;
```
this could do with some comments on usage. Is this taken from GDB ABI ?

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+	struct device *dev = &cdev->auxdev.dev;
+	int ret, err;
+
+	ret = pm_runtime_resume_and_get(dev);
```
If we are debugging do we really want to support suspend/resume. If we are debugging a SOF module we should force device into D0.

lyakh replied:
@lgirdwood This is fixed in the fourth patch. I tried to keep the first commit as close to the original as possible, at least functionally. But I think I'll address at least all the comment style and lack of comment comments in it directly, instead of adding follow-up patches

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+		return -ENODEV;
+
+	ret = debugfs_file_get(file->f_path.dentry);
+	if (unlikely(ret))
```
unlikely is overkill here, not a high frequency call.

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+	if (ret)
+		debugfs_file_put(file->f_path.dentry);
+
+	// When file is opened, send GDB init command to FW
```
C style.

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+	int count = 0;
+	int bar = snd_sof_dsp_get_bar_index(sdev, SOF_FW_BLK_TYPE_SRAM);
+	void __iomem *rx_ring_buffer = sdev->bar[bar] + sdev->debug_box.offset + 256;
+	char* msg_region = rx_ring_buffer + 128;
```
@ranj063 don we have some API or core context that has this ?

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+
+
+	uint8_t head = readb(rx_ring_buffer);
+	uint8_t tail = readb(rx_ring_buffer + 64);
```
magic numbers

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+	uint8_t head = readb(rx_ring_buffer);
+	uint8_t tail = readb(rx_ring_buffer + 64);
+
+	for (int i = 0; i < bytes; i++) {
```
lets comment this.

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+
+	for (int i = 0; i < bytes; i++) {
+		if (tail == head) {
+			// No Data
```
C style

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+		count++;
+	}
+
+	writeb(tail, rx_ring_buffer + 64);
```
magic

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+	return count;
+}
+
+static ssize_t sof_fw_gdb_dfs_read(struct file *file, char __user *buffer,
```
we need a lot of comments explaining how we tunnel GDB commands to IPC4 and vice versa.

lyakh replied:
@lgirdwood now I see what tunnel you mean. In fact it's simple - GDB message strings are just copied 1-to-1 into memory windows, practically zero tunnelling.

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+	return count_read;
+}
+
+// TODO, should be same as ipc3
```
can be added by ipc3 users.

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+	priv->dfs_file = debugfs_create_file("fw_gdb", 0644, debugfs_root,
+					     cdev, fops);
+
+	/* enable runtime PM */
```
lets disable runtime PM

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,333 @@
+module_auxiliary_driver(sof_gdb_fw_client_drv);
+
+MODULE_DESCRIPTION("SOF IPC FW GDB Client Driver");
+MODULE_LICENSE("GPL");
```
dual license with BSD.

lgirdwood commented on sound/soc/sof/Kconfig:
```diff
@@ -258,6 +258,10 @@ config SND_SOC_SOF_DEBUG_IPC_KERNEL_INJECTOR
 	  Say Y if you want to enable the IPC kernel injector.
 	  If unsure, select "N".
 
+config SND_SOC_SOF_DEBUG_FW_GDB
+	tristate "SOF enable Firmware GDB debugging"
+	select SND_SOC_SOF_CLIENT
+
```
In general I think this file needs some refinement around the tunnel protocol handling.

lyakh replied:
@lgirdwood sorry, which tunnel protocol do you mean?

lgirdwood replied:
I mean the copying/handling of GDB data to/from IPC layer, but I think you've already done this in the later patches in this PR.

lyakh replied:
@lgirdwood gdb communication isn't done via IPCs, it's done via a memory window. Only the first "switch to gdb" command is an IPC, and that isn't a part of the actual gdb messaging yet.

lgirdwood replied:
ok, but then how would stop/break work if there is no IRQ ?

lyakh replied:
@lgirdwood that's what the added polling periodic delayed work does in https://github.com/thesofproject/linux/pull/5411/commits/551692fe8ac266ecd90efbeb462381c9f29f6142#diff-fdf6d136ee82a240596bbe084669793c8f1ca4499e85a85e92dba039fd81461aR310 . And it isn't only needed for breaks, it's needed for any replies to gdb commands

lgirdwood commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -22,36 +24,57 @@
+		struct sof_ipc4_msg msg;
+		char cmd[];
+	} ipc4;
+} __packed;
```
looks like this can be squashed.

lyakh replied:
@lgirdwood You mean as
```
struct sof_ipc_gdb_message {
    union {
        struct sof_ipc_cmd_hdr hdr;
        struct sof_ipc4_msg msg;
    };
    char cmd[];
} __packed;
````
? Yeah, I'll double check if this works and if yes, will use it, thanks

lgirdwood replied:
sorry, I meant you are adding a new file to the kernel in a series of patches. You should just do it a 1 patch for the new file.

lyakh replied:
@lgirdwood well, I tried to keep the first patch as close to the author's original as possible. But if we decide that that isn't needed, sure, can squash

lgirdwood reviewed (commented):
Stopped reviewing as it looks like some new files can be squashed.

lyakh committed:
ASoC: SOF: ipc4: add a command to start GDB

Add an IPC4 message type to start a firmware GDB debugging session.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,445 @@
+	struct sof_fw_gdb_priv *priv = cdev->data;
+	struct snd_sof_dev *sdev = sof_client_dev_to_sof_dev(cdev);
+	char *kbuff = kzalloc(count + 1, GFP_KERNEL);
+	unsigned long num_failed = copy_from_user(kbuff, buffer, count);
```
kbuff might be NULL at this point...

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,445 @@
+		return -EIO;
+	}
+
+	for (int i = 0; i < bytes; i++) {
```
we usually don't do this in kernel C code...

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,445 @@
+	char* msg_region = tx_ring_buffer + SOF_GDB_HEADER_SIZE;
+
+	if (head >= RING_SIZE) {
+		dev_err(sdev->dev, "%s(): %p head %x beyond buffer limit!\n",
```
you can drop () from the prints, it is not aligning with the style used in kernel and SOF.

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,445 @@
+
+	mutex_unlock(&priv->mutex);
+
+	dev_dbg(&cdev->auxdev.dev, "debug send %s\n", kbuff);
```
what does 'debug send' means?

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,478 @@
+
+		msg->primary = SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_GLB_ENTER_GDB) |
+			SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST) |
+			SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
```
nitpick: you can align the start of the lines here

lyakh replied:
I've looked through a couple of SOF files and haven't found any similar broken lines, so I'd rather just leave it at my editor's choice :-)

ujfalusi replied:
Right, because they are implemented this way always:
```c
msg->primary = SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_LARGE_CONFIG_SET);
msg->primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
msg->primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG);
```

lyakh replied:
@ujfalusi exactly, which I also dislike :-) To me it looks like
```
a = 2;
a = a + 3;
```
instead of writing
```
a = 2 + 3;
```
but let's say it's subjective :-)

ujfalusi replied:
I cannot say why it is like that, but clearly it is followed throughout sof kernel code...

lyakh replied:
> I cannot say why it is like that, but clearly it is followed throughout sof kernel code...

@ujfalusi the entire kernel? Not only SOF / ASoC? I'm sure the kernel has enough examples of both, but counting them wouldn't be very simple...

ujfalusi replied:
> > I cannot say why it is like that, but clearly it is followed throughout sof kernel code...
> 
> @ujfalusi the entire kernel? Not only SOF / ASoC? I'm sure the kernel has enough examples of both, but counting them wouldn't be very simple...

I cannot say why it is like that, but clearly it is followed throughout **sof** kernel code...

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,478 @@
+	switch (priv->ipc_type) {
+	case SOF_IPC_TYPE_3:
+		;
+		struct sof_ipc_cmd_hdr *hdr = &priv->tx_buffer->ipc3.hdr;
```
In other parts of the SOF code we use {} when we use local variables in switch:
```c

	case SOF_IPC_TYPE_3:
	{
		struct sof_ipc_cmd_hdr *hdr = &priv->tx_buffer->ipc3.hdr;

		hdr->cmd = SOF_IPC_GLB_GDB_DEBUG;
		hdr->size = sizeof(*hdr);
		break;
	}
```

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,478 @@
+	struct sof_fw_gdb_priv *priv = cdev->data;
+	struct snd_sof_dev *sdev = sof_client_dev_to_sof_dev(cdev);
+	char *kbuff = kzalloc(count, GFP_KERNEL);
+
```
no need for blank line

lyakh replied:
this would be my personal preference too, but some style checkers complain about a lack of an empty line between "declarations and code." So I forced myself to adapt this style...

ujfalusi replied:
Oh yes, it is correct with the blank line, sorry.

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,478 @@
+	unsigned long remain = copy_to_user(buffer, kbuff, count_read);
+
+	kfree(kbuff);
+
```
extra line is not needed.

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,478 @@
+		gdb_is_open = true;
+	}
+
+	priv->sdev = sof_client_dev_to_sof_dev(cdev);
```
sof-clients are restricted to use sdev directly.

ujfalusi replied:
I think I will rework the sof-client.h to make this more evident (header for client driver use only and rest which is needed by the core).

lyakh replied:
> sof-clients are restricted to use sdev directly.

@ujfalusi do you mean clients should only call accessor functions to access well defined sdev functionality? For GDB we need this:
```
	int bar = snd_sof_dsp_get_bar_index(sdev, SOF_FW_BLK_TYPE_SRAM);
	void __iomem *ring_buffer = sdev->bar[bar] + sdev->debug_box.offset +
		SOF_GDB_WINDOW_OFFSET + offset;
```
Can we get this via those accessors? But the client API has `sof_client_dev_to_sof_dev()` - should it not be available at all?

lyakh replied:
...or do you mean that clients just shouldn't save the pointer locally and should always use `sof_client_dev_to_sof_dev()` instead and save the client pointer?

ujfalusi replied:
What I mean is that clients must not have any knowledge of `sdev`, in your case you should ask for a debug slot by index as you have re-purposing the descriptor slot with second functionality.
We support maximum of 15 slots in the debug windows:
```
------------------------
| Page0  - descriptors |
------------------------
| Page1  - slot0       |
------------------------
| Page2  - slot1       |
------------------------
|         ...          |
------------------------
| Page14  - slot13     |
------------------------
| Page15  - slot14     |
------------------------
 ```
we have maximum of 14 descriptors for the slots, but there is no descriptor for self, so you cannot say that page0 is descriptor _and_ GDB, which is a bummer. The slots themself does not store their magic ID either.

But what if we introduce a descriptor15 which is a special one describing what is using the page0's free part?

You could then ask for the slot by ID `0x42444700` and get the offset, but hrm, the size then would not be less that the page size... But you can define that the GDB slot size is always this not full slot, even if it is occupying a dedicated one.

ujfalusi replied:
> ...or do you mean that clients just shouldn't save the pointer locally and should always use `sof_client_dev_to_sof_dev()` instead and save the client pointer?

Client drivers in _their_ code must not use anything sdev, they must use wrappers or APIs and yes, I will try to hide this from the client drivers somehow. I have it in the header so we can have light inline code as API, but it is prone to abuse.

ujfalusi replied:
@lyakh, another possibility is to have a way to request a slot by page index, but I think the issue is: how can you know that the GDB is using that slot and not another, is the GDB enabled in the first place? You need some way to tell that you can go ahead and use the are and not just assume that it will be always there

lyakh replied:
@ujfalusi yes, we just need to pick up something... @kv2019i @lgirdwood what do you think?

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,478 @@
+
+	priv->sdev = sof_client_dev_to_sof_dev(cdev);
+
+	schedule_delayed_work(&priv->poll_work, msecs_to_jiffies(SOF_FS_GDB_POLL_DELAY_MS));
```
can the GDM send IPC notification? We have a way to register for them in clients.

lyakh replied:
@ujfalusi no, don't think you should use the IPC sending machinery in the GDB stub. When the GDB stub is running, the SOF is *halted* at some random location, so you shouldn't use any infrastructure that is also used by SOF.

ujfalusi replied:
@lyakh, OK, I have no idea what the GDB stub does to be honest..

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,478 @@
+{
+	struct sof_client_dev *cdev = file->private_data;
+	struct sof_fw_gdb_priv *priv = cdev->data;
+	struct snd_sof_dev *sdev = sof_client_dev_to_sof_dev(cdev);
```
while sdev is not allowed to be used by clients directly, you have saved this to priv->sdev in sof_fw_gdb_dfs_open()

lyakh replied:
AFAIU clients do use `sdev`, it does seem to be allowed and a part of the SOF client API. You seem to just prefer not to let clients access *internal* structure of `sdev`, which I'm doing in the updated version. I could of course save `cdev` instead of `sdev` in private data, but I anyway need `sdev` in the end.

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,478 @@
+{
+	int bar = snd_sof_dsp_get_bar_index(sdev, SOF_FW_BLK_TYPE_SRAM);
+	void __iomem *ring_buffer = sdev->bar[bar] + sdev->debug_box.offset +
+		SOF_GDB_WINDOW_OFFSET + offset;
```
The GDB is part of the debug window? How does it co-exists with other debug window users? How the slot is allocated? Is it true  that this magical offset is _always_ GDB slot and never can be anything else?

We have this for IPC4:
```c
include/sound/sof/ipc4/header.h:#define SOF_IPC4_DEBUG_SLOT_GDB_STUB            0x42444700
```
and we have (in core) this to look up a slot: sof_ipc4_find_debug_slot_offset_by_type(), you would need a client wrapper for this and use it to make sure that the assumption is firing back.

lyakh replied:
@ujfalusi the problem is that we don't have many free debug slots, in fact we have none. Debug slots are page-size parts of the second memory window. The first page (slot 0) ATM is used for slot descriptors... Whereas those descriptors use a few hundreds of bytes - less than 1kB. Slot 1 is used for logging. On TGL that's it - window 2 is only 8kB (2 page) large. On other platforms like MTL we have generous 12kB there, i.e. 3 pages / 3 slots. Slot 2 is reserved for the shell. So, we can either say, that shell and GDB are mutually exclusive, but that still wouldn't solve the problem for TGL. And on PTL it's 2 pages again.
So, what I've done - I grabbed the remaining 3kB of slot 0. Nobody needs them, nobody uses them, but we need to guard that well...

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,478 @@
+		return -ENOMEM;
+
+	priv->ipc_type = sof_client_get_ipc_type(cdev);
+	priv->max_msg_size = sizeof(*priv->tx_buffer);
```
is this right?
this is the size of the `union sof_ipc_gdb_message` struct. You are never sending/receiving additonal data on top of a headers?

lyakh replied:
@ujfalusi ATM we only have 1 IPC - switch to GDB stub. And it doesn't have any data. So in principle we could remove those data pointers. If we need data in the future - we can add those and it won't break the ABI, because there's anyway data size included with IPCs, right?

ujfalusi replied:
not quite, in ipc4 the data size is what you have placing into the mailbox, probably you want to use sof_client_ipc_tx_message_no_reply() here and have not reply specified at all.

lyakh replied:
> not quite, in ipc4 the data size is what you have placing into the mailbox, probably you want to use sof_client_ipc_tx_message_no_reply() here and have not reply specified at all.

that's much better, thanks!

ujfalusi reviewed (changes requested):
There are two big issues I can see:
- client drivers are restricted from using `sdev` directly.
- the debug slot assumption is dangerous, what happens if you don't even have GDB support in fw and thus no GDB slot?

lyakh committed:
ASoC: SOF: special-case a part of the debug memory window

The debug memory window is split in 4KiB large pages. Page zero is
used for slot descriptors, which leaves one or two more free slots
for actual useful work. One of them is usually used for mtrace
logging, which leaves no free slots on some systems and only one such
slot on others. This commit repurposes the free space in page zero
after slot descriptors to be used as slot number 16, one higher than
the currently highest possible slot number. To improve documentation
we move the tabular representation of the window layout from
ipc4-mtrace.c to header.h.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
ASoC: SOF: client: add wrappers for memory window access

Add wrappers for mailbox reading and writing and for debug slot
offset extraction.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

ujfalusi commented on sound/soc/sof/ipc4.c:
```diff
@@ -149,6 +149,7 @@ static const char * const ipc4_dbg_glb_msg_type[] = {
 	DBG_IPC4_MSG_TYPE_ENTRY(GLB_BOOT_CONFIG),
 	DBG_IPC4_MSG_TYPE_ENTRY(GLB_ROM_CONTROL),
 	DBG_IPC4_MSG_TYPE_ENTRY(GLB_IPCGATEWAY_CMD),
+	DBG_IPC4_MSG_TYPE_ENTRY(GLB_ENTER_GDB),
```
Keep this list ordered by index, so the GDB goes to the end of the list, after SOF_IPC4_GLB_NOTIFICATION

ujfalusi commented on sound/soc/sof/ipc4.c:
```diff
@@ -577,6 +577,21 @@ static int sof_ipc4_init_msg_memory(struct snd_sof_dev *sdev)
+		return sdev->debug_box.offset + index * SOF_IPC4_DEBUG_SLOT_SIZE;
+
+	/* Slot 15: see comment for SOF_IPC4_MAX_DEBUG_SLOTS in header.h */
+	return sdev->debug_box.offset + 1024;
```
How do we know that GDM is in slot0?

I still think it would be better to look up the offset by type for GDM as well.
descriptor 15 (a new one since atm we have 14 of them) would be used to describe the remaining of slot0:
```
desc0 -> slot1
desc1 -> slot2
..
desc13 -> slot14
desc14 -> slot15
desc15 -> slot0+1024
```
in sof_ipc4_find_debug_slot_offset_by_type() after the for loop, check if the desc15 has the match
```c
sof_mailbox_read(sdev, slot_desc_type_offset, &type, sizeof(type));
if (type == slot_type)
	return sdev->debug_box.offset + 1024;

dev_dbg(sdev->dev, "Slot type %#x is not available in debug window\n", slot_type);
return 0;
```

And in GDB client driver:
```c
offset = sof_client_ipc4_debug_slot_offset_by_type(cdev, SOF_IPC4_DEBUG_SLOT_GDB_STUB);
size = SOF_IPC4_DEBUG_SLOT_SIZE - (offset & (SOF_IPC4_DEBUG_SLOT_SIZE - 1));
```
this would make things forward compatible in regards where GDB is and we will also be able to detect that it is available at all via the debug window.

ujfalusi commented on sound/soc/sof/sof-client.c:
```diff
@@ -673,3 +673,24 @@ struct snd_sof_dev *sof_client_dev_to_sof_dev(struct sof_client_dev *cdev)
+{
+	return sof_ipc4_find_debug_slot_offset_by_index(sof_client_dev_to_sof_dev(cdev), index);
+}
+EXPORT_SYMBOL_NS_GPL(sof_client_ipc4_find_debug_slot_offset_by_index, "SND_SOC_SOF_CLIENT");
```
See my previous comment on using the lookup by type.

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,482 @@
+			SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
+		msg->extension = 0;
+		msg->data_size = 0;
+		msg->data_ptr = NULL;
```
These are 0 after the allocation and are not changed during runtime, no need to clear them?

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,482 @@
+									SOF_IPC4_MAX_DEBUG_SLOTS);
+
+	if (slot_offset < 0)
+		return slot_offset;
```
offset == 0 is an error.

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,482 @@
+
+	/* When file is opened, send GDB init command to FW */
+	if (!gdb_is_open) {
+		ret = sof_fw_gdb_send_message(cdev);
```
If you _need_ the firmware to be booted up when the gdb is opened for the first time you will need to call `sof_client_boot_dsp()` after the runtime resume.

ujfalusi reviewed (commented):
@lyakh, I think the code looks mostly good, but I still think that we should look up the GDB slot by type, see my comments.

ujfalusi commented on include/sound/sof/ipc4/header.h:
```diff
@@ -544,11 +544,19 @@ struct sof_ipc4_notify_resource_data {
+#define SOF_IPC4_MAX_DEBUG_SLOTS	15				/* usable slots 1..15 */
+#define SOF_IPC4_MAX_DEBUG_DESCS	(SOF_IPC4_MAX_DEBUG_SLOTS + 1)	/* including descriptor
+									 * for partial slot 0 */
+#define SOF_IPC4_DEBUG_SLOT_SIZE	0x1000
```
These were aligned with the debug slot type defines and while we are here I think we should correct the definitions and terms:
```c
/*
 * debug info window is organized in 16 (equal sized) pages:
 *
 *	------------------------
 *	| Page0  - descriptors |
 *	------------------------
 *	| Page1  - slot0       |
 *	------------------------
 *	| Page2  - slot1       |
 *	------------------------
 *	|         ...          |
 *	------------------------
 *	| Page14  - slot13     |
 *	------------------------
 *	| Page15  - slot14     |
 *	------------------------
 *
 * The slot size == page size
 */
#define SOF_IPC4_DEBUG_PAGE_SIZE		0x1000
#define SOF_IPC4_DEBUG_SLOT_SIZE		SOF_IPC4_DEBUG_PAGE_SIZE
#define SOF_IPC4_DEBUG_PAGE0_SLOT_OFFSET	1024

/* usable slots 0..14 (page 1..15) */
#define SOF_IPC4_MAX_DEBUG_SLOTS		15

/*
 * Descriptor 0-14: slot 0-14 (page 1-15)
 * Descriptor 15: partial slot at page 0 + 1024
 */
#define SOF_IPC4_MAX_DEBUG_DESCS		(SOF_IPC4_MAX_DEBUG_SLOTS + 1)
#define SOF_IPC4_DEBUG_DESCRIPTOR_SIZE		12 /* 3 x u32 */

/* debug log slot types */
```

ujfalusi commented on sound/soc/sof/ipc4.c:
```diff
@@ -586,11 +586,16 @@ size_t sof_ipc4_find_debug_slot_offset_by_type(struct snd_sof_dev *sdev,
+				return sdev->debug_box.offset + 1024;
+
 			return sdev->debug_box.offset + (i + 1) * SOF_IPC4_DEBUG_SLOT_SIZE;
+		}
```
I think it would be a bit cleaner if you check the descriptor 15 after the loop, which checks the standard slots:
```c
for (i = 0; i < SOF_IPC4_MAX_DEBUG_SLOTS; i++) {
...
}

/* Check the descriptor 15 for the partial slot at page0 */
sof_mailbox_read(sdev, slot_desc_type_offset, &type, sizeof(type));
if (type == slot_type)
	return sdev->debug_box.offset + SOF_IPC4_DEBUG_PAGE0_SLOT_OFFSET;

dev_dbg(sdev->dev, "Slot type %#x is not available in debug window\n", slot_type);
```

lyakh replied:
@ujfalusi sorry, I'd rather avoid redundancy: we already have `sof_mailbox_read()` and `if (type == slot_type)` in the loop, don't think we should take them out. Also you wanted to have this "partial" slot to see as normal as possible - include it into the descriptor list etc. So special casing it here would contradict that effort IMHO.

ujfalusi replied:
OK

ujfalusi commented on sound/soc/sof/sof-client-fw-gdb.c:
```diff
@@ -0,0 +1,485 @@
+ * compatibility but we should use a free slot on other platforms and be careful
+ * on TGL
+ */
+#define SOF_GDB_WINDOW_OFFSET 1024
```
this define is no longer needed

ujfalusi reviewed (commented):
@lyakh, I think this looks great now, just few things I would like to ask to be changed and improved.

ujfalusi commented on include/sound/sof/ipc4/header.h:
```diff
@@ -541,16 +541,45 @@ struct sof_ipc4_notify_resource_data {
+ * u32 vma;
  */
+
+#define SOF_IPC4_DEBUG_PAGE_SIZE		(4 * 1024)
```
what was wrong with 0x1000, which is btw used also in firmware.

lyakh replied:
@ujfalusi having 0x1000 used in one location and 1024 in another was weird. So I decided that `(4 * 1024)` is easier than `0x400`

ujfalusi replied:
I prefer 0x1000 for the PAGE and SLOT size and I'm indifferent on the offset representation within PAGE0

ujfalusi reviewed (commented):
@lyakh , looks all good except the obfuscation of the SOF_IPC4_DEBUG_PAGE_SIZE value, with that reverted to 0x1000 I'm ready to approve this PR.

lyakh commented:
3 latest minor changes:
* https://github.com/thesofproject/linux/compare/38e32d0487092a0ea58674fe57731e7d36ab9ba6..3b29d400135085399712be77424fbb803b2f9ab1 - changed back to `0x1000`
* https://github.com/thesofproject/linux/compare/3b29d400135085399712be77424fbb803b2f9ab1..865fdf77e5e95af18e5f36dd6fc8667efbd767d5 - moved `sof_client_ipc4_find_debug_slot_offset_by_type()` under `#if IPC4` to fix IPC3 builds
* https://github.com/thesofproject/linux/compare/3b29d400135085399712be77424fbb803b2f9ab1..865fdf77e5e95af18e5f36dd6fc8667efbd767d5 - added checks for `sof_fw_gdb_get_tx_ptrs()` return code to make sure not to use `head` and `tail` if it fails

bardliao commented:
@lyakh Could you take a look at the spare check issue below?
```
CHECK   sound/soc/sof/sof-client-fw-gdb.c
sound/soc/sof/sof-client-fw-gdb.c:365:22: error: invalid assignment: |=
sound/soc/sof/sof-client-fw-gdb.c:365:22:    left side has type restricted __poll_t
sound/soc/sof/sof-client-fw-gdb.c:365:22:    right side has type int
sound/soc/sof/sof-client-fw-gdb.c:369:22: error: invalid assignment: |=
sound/soc/sof/sof-client-fw-gdb.c:369:22:    left side has type restricted __poll_t
sound/soc/sof/sof-client-fw-gdb.c:369:22:    right side has type int
make[5]: *** [scripts/Makefile.build:288: sound/soc/sof/sof-client-fw-gdb.o] Error 1
```
